### DISCUSSION TITLE
JS Proxy on Clr Object

### DISCUSSION THREAD

**Discussion by martinbu** [sebastienros/jint#1644] [Q&A] [answered]
Hi,
I am using Jint 3.0.0-beta-2052. I am testing the Js Proxy where I get exceptions if I access a property that exists on a clr object that the proxy was created on. If the property does not exist on the clr object, everything works as expected.
Also everything works fine if the proxy is created on a class created in JS instead of a clr class.

Example

```C#
class TestClass
{
     public string Name => "My Name is Test";
}

```

```C#
new Engine()
    .SetValue("T", new TestClass())
    .Execute("""
        const handler = {
            get(target, property, receiver) {

                if (!target[property]) {
                    return (...args) => "Not available";
                }

                // return target[property];
                return Reflect.get(target, property, receiver);
            }
        };

        const p = new Proxy(T, handler);
        const t = p.GetX();
        // const name = p.Name; // When commenting this in, an exception is thrown
    """);
```

Am I doing something wrong here or did I understand something wrong? Or isn't it supported to use a Proxy on a clr object?

**Comment by viruscamp** [comment]
I have made a mistake below, It's really a bug.

~~It's the right behavior in [ECMAScript Language Specification](https://tc39.es/ecma262/multipage/ordinary-and-exotic-objects-behaviours.html#sec-proxy-object-internal-methods-and-internal-slots-get-p-receiver).~~  
See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy/Proxy/get#invariants   
>The value reported for a property must be the same as the value of the corresponding target object property if the target object property is a non-writable, non-configurable own data property.

All CLR Object Property is non-configurable, and the `Name` property is obviously non-writable in your code.

**Comment by viruscamp** [comment]
It's all in `JsProxy.Get`:  
1. Error message must be improved here
2. It called CLR getter twice in `TryCallHandler`
3. Compared the value in `ReferenceEquals` will call CLR getter once more
4. Each call to CLR getter returns a new `JsValue`, so they cannot be `ReferenceEquals`ed

```c#
        /// <summary>
        /// https://tc39.es/ecma262/#sec-proxy-object-internal-methods-and-internal-slots-get-p-receiver
        /// </summary>
        public override JsValue Get(JsValue property, JsValue receiver)
        {
            AssertTargetNotRevoked(property);
            var target = _target;

            if (property == KeyFunctionRevoke || !TryCallHandler(TrapGet, new[] { target, TypeConverter.ToPropertyKey(property), receiver }, out var result))
            {
                return target.Get(property, receiver);
            }

            var targetDesc = target.GetOwnProperty(property);
            if (targetDesc != PropertyDescriptor.Undefined)
            {
                var targetValue = targetDesc.Value;
                if (targetDesc.IsDataDescriptor() && !targetDesc.Configurable && !targetDesc.Writable && !ReferenceEquals(result, targetValue))
                {
                    ExceptionHelper.ThrowTypeError(_engine.Realm, $"'get' on proxy: property '{property}' is a read-only and non-configurable data property on the proxy target but the proxy did not return its actual value (expected '{targetValue}' but got '{result}')");
                }
                if (targetDesc.IsAccessorDescriptor() && !targetDesc.Configurable && (targetDesc.Get ?? Undefined).IsUndefined() && !result.IsUndefined())
                {
                    ExceptionHelper.ThrowTypeError(_engine.Realm, $"'get' on proxy: property '{property}' is a non-configurable accessor property on the proxy target and does not have a getter function, but the trap did not return 'undefined' (got '{result}')");
                }
            }

            return result;
        }
```

**Reply by martinbu** [reply]
Thank you for investigation @viruscamp
Should the issue then be reopended?

**Comment by viruscamp** [comment, marked as answer]
@lahma Please reopen #1643 , I can comfirm it's a bug and I can provide a fix:  
1. provide detailed message for exceptions in JsProxy
2. donot use `ReferenceEquals` on data property value comparing
3. maybe we should mark all Clr Property and Indexer as js accessor property

That's a pure js example which works in V8, but failed in Jint, caused by  `ReferenceEquals` 
```javascript
            let o = Object.defineProperty({}, 'value', {
              configurable: false,
              value: 'oz',
            });
            const handler = {
              get(target, property, receiver) {
                return 'Mozilla'.substring(1,3);
              }
            };
            let p = new Proxy(o, handler);
            let pv = p.value;
```

**Comment by martinbu** [comment]
I can confirm that now the example works as expected. Thank you!
